Fix MQTT double topic subscribtion

corporategoth opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Right now, the MQTT integration will attempt to subscribe to a topic multiple times if that topic is used by multiple entities.  Either for availability or data.

This will ensure that we only subscribe once, and will just duplicate the messages received as necessary.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #87965, #87972
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

corporategoth commented:
There may still be a minor issue with this.  It works perfectly, but only after I use the configure / listen option.  After which it's all fine (even when I stop listening). However, this DOES actually work - where as previously this issue would mean entities did not get configured correctly and startup would take forever.

I will investigate this minor issue tomorrow.  But this, or something like this needs to be in place for large MQTT (or zigbee2mqtt) setups.

bdraco commented:
This needs to be fixed in the `MQTT` client itself

There is already code to prevent dupe unsub https://github.com/home-assistant/core/blob/f11ba1b7cb6107bd0f5aa27fb8935eebd84a1905/homeassistant/components/mqtt/client.py#L547

https://github.com/home-assistant/core/blob/dev/homeassistant/components/mqtt/client.py#L486
https://github.com/home-assistant/core/blob/dev/homeassistant/components/mqtt/client.py#L535

There needs to be a lock though since there is currently a race where you can have multiple calls to subscribe on the same topic

Also the subscriptions needed to be indexed

corporategoth commented:
*nod* I'll look into moving the code into MQTT itself.  TBH, I was working on this in the very wee hours.

jbouwh commented:
> There needs to be a lock though since there is currently a race where you can have multiple calls to subscribe on the same topic
> 
> Also the subscriptions needed to be indexed

https://github.com/home-assistant/core/pull/88117 Will fix the lock issue.

jbouwh reviewed (changes requested):
Please run local tests first and make sure the pass.
Run mypy and apply correct strict typing.
Run black and other linters.
Add tests.

corporategoth committed:
Updated implementation

corporategoth committed:
Updated to use _is_active_subscription call

corporategoth committed:
Fixed typo.

jbouwh commented:
Also have a look at this test:
https://github.com/home-assistant/core/pull/88138/files#diff-466271ea1d386b60cc3ec712abd4078d0260097e8d4e9ed79b397a4974e60634R1288-R1322

corporategoth committed:
Updated unit tests.
PyTest and pre-commit hooks all work now.

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -515,8 +517,16 @@ async def async_subscribe(
+            if already_active_subscription:
+                _LOGGER.debug("Skipped duplicate subscribe for %s", topic)
+                if topic in self._last_message:
+                    self.hass.async_run_hass_job(
+                        subscription.job, self._last_message.get(topic)
+                    )
+                    self._mqtt_data.state_write_requests.process_write_state_requests()
```
This is not correct, it means we treat non-retained MQTT messages as if they are retained.

corporategoth replied:
> It would be great if we could avoid subscribing to the same topic multiple times

It's not just great.  It's necessary, this resolves two major issues.

First, is the absolute flood of messages at start up caused by multiple subscribes. In my instance, that's more than 8000 subscribe requests to a single topic (the zigbee2mqtt availability topic) - and that's on TOP of the multiple subscriptions for each entity.  This causes the startup of HASS to be more than 10 minutes, and many of these requests time out or don't get processed (or worse, re-processed for every entity for each subscription).
Second, related, is that the subscriptions are going out so much we are (I assume) getting throttled?  Not sure, but many of the subscription requests don't actually return anything.  So we don't actually get subscribed for many entities, meaning those entities aren't listening on the topics they need to (despite the MQTT being subscribed to it), and therefore don't get updates.

Without this patch, HA MQTT is completely broken.  And I don't mean that as hyperbole.  I mean, in my instance, with only 100 or so Zigbee devices feeding into MQTT (Which ends up being more than 8000 entities, for some reason - though about half of them I think are inactive/should be removed because they were renamed - so let's say 4000) - the MAJORITY of the entities NEVER get updates due to this patch because of the subscribe calls being too many and never returning.  And startup takes 10 minutes waiting for MQTT to issue all these duplicate subscribes.

> When sending a PUBLISH Packet to a Client the Server MUST set the RETAIN flag to 1 if a message is sent as a result of a new
> subscription being made by a Client [MQTT-3.3.1-8]. It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to a Client because it matches an established subscription **regardless of how the flag was set in the message it received** [MQTT-3.3.1-9].

The part of the document you pasted explicitly says the server is dropping and ignoring the value of the PUBLISH flag it received from the client - and setting it based on whether the message is sent on an existing or new subscription.  Therefore we don't NEED to worry about what the original publisher of the message set the RETAIN flag.  I added emphasis to show this.

By my reading of this, this is the flow.

Client A subscribes to foo - get message with the RETAIN flag set to 1
Client B subscribes to foo, client A and client B both get a message, but for client A the RETAIN flag is set to 0, and for client B it's set to 1.
Client C subscribes to foo, client A, B and C get a message, but for clients A and B the RETAIN flag is set to 0, and for client C it's set to 1.
This is all true regardless of what the publisher of the original message set the RETAIN flag.

In this case, I think we are fine either way.  I would say that we should consider HA MQTT the client, not every individual entity HA MQTT provides to HASS.  Meaning that, with this, HA MQTT subscribes once, and if something else (something other than HASS) subscribed to those topics, we would see messages with RETAIN=0 when they do.

This makes the most sense - since @jbouwh already said that, when we send a new subscribe message, we are not subscribing a second time - MQTT will only allow one subscription per client per topic - and client in this case is indicated as a single socket connection between HA MQTT and the MQTT server.

Since we don't create/configure all entities first, THEN subscribe to the topics for each entity, but instead subscribe when we first see the entity, we are just informing each new "subscriber" of the current data for that topic.  From a MQTT perspective, each entity is **not** a new client, and we should not be re-subscribing - something that MQTT will not allow any (one subscription per client) - we are just needlessly flooding the MQTT server with subscribe requests, and ourselves with duplicate responses.

It's not like we'd miss updates - once we're subscribed, we will be getting the updated information reardless.

corporategoth replied:
So I guess the exception to this is:

- Client sends message A with RETAIN 1
- MQTT sends message A to all other clients (with retain set above)
- Client sends message B with RETAIN 0
- MQTT sends message B to all other clients (with retain set above)
- New entity subscribes to topic, and should receive message A (not B) because THAT was what was the retained message in MQTT.  However, with this patch, they will receive message B.


There is no way we can get around this, that's true, as we don't know what the original RETAIN value was - but I would argue that this is not a frequent case, and more importantly, if we DON'T implement this change, the HA MQTT server is unusable for people who have a high volume of topics or high volume of entities subscribed to the same topic (which is the case for Zigbee2MQTT users, as EVERY entity created subscribes to the same status topic).

Even the CURRENT behavior we use is wrong.

Right now, this is the current behavior - without this PR:

- External client sends message A with RETAIN=1
- We subscribe a bunch of times, for new entities.  Each time we subscribe, EVERY entity we have subscribed so far gets message A. So if we subscribe 5 entities, the first entity gets message 1 5 times, the next 4 times, the next 3 times, etc.
- External client sends message B with RETAIN=0
- All 5 of our subscribed entities get message B
- We subscribe a sixth entity, and then all 6 entities now get blasted with message A again.

We should never have sent message A to the first 5 entities that were already subsribed with the subscription of entity 6, they already got message B (ephemeral as it was), and had already processed message A.  But now they will all be re-processing the retained message, despite the fact that, if they were all truly separate MQTT clients (as IT defines client), they would never have been re-issued a message A publish.


There are two potential ways around this issue:
1. We delay ALL subscriptions until discovery is complete.  We create the internal subscriptions, so that when we DO do the MQTT subscription, they will all get the messages, but we don't actually subscribe until we have completed processing of the initial blast of discovery messages.
2. Every N seconds, allow a new subscribe message to go through on a new subscription, which would trigger sending message A to everyone who "fake subscribed" and got message B (but not those who already got 'A'.

The drawback to option 1 is that it's a much bigger solution, and I'd rather fix the completely broken state of HA-MQTT for high-entity clients than wait on that.  And possibly have doing that as a future task.  Plus it means we won't get any data for any entity until discovery and entity creation is over (not necessarily a bad thing).  That said though, we have no real way of knowing when discovery is over.
The drawback for option 2 is it's  hack, and not reliable.

corporategoth replied:
FYI - here is what the code to not spam previously subscribed topics looks like
https://github.com/corporategoth/core/pull/2

But I don't think this is the correct way to go - because although it DOES fix the extremely slow startup problem (not re-processing the initial messages for all new subscreiptions) - it doesn't fix the spamming MQTT repeatedly for the same topic problem.  And this problem is what is causing a large percentage of my existing entities to fail - the subscription request for them is dropped on the floor (either by MQTT or HA or something).

bdraco replied:
Subscribing to the same topic is effectively O(n²) performant. That's not great, as it doesn't scale well.

As this PR stands now, it's a significant improvement for one set of users. However, there is a risk of collateral damage, and we are trading one problem for another, which may create a fire for the maintainers of this integration to put out, making everyone reluctant to want to merge this in the current state.

You have put a lot of thought into this. Please find a solution to address Erik's feedback above so we can merge this.

corporategoth replied:
@bdraco I mentioned 3 different "solutions" - along with the drawbacks for each, addressing Erik's feedback.  And have not gotten feedback on any one of them.

> "Please solve P=NP before we will think about patching the leak in our boat making it unseaworthy."
> "Here are some potential problems with drawbacks, what are your thoughts?"
> "Solve P=NP before we fix the boat!"
> "That does not give me any information to go on."
> "Just come up with a solution!  I'm not going to help you!  Use magic to get the perfect solution!  We won't accept anything less.  But hurry, because we really WOULD like to fix this leaky boat."
> "..."

I submitted this MR to solve an emergent (and debilitating to the point of the MQTT being broken at any reasonable scale) problem, mainly because @jbouwh kept closing my bug reports as "not going to fix" or "probably zigbee2mqtts problem".  I wasn't planning on re-designing how the mqtt module works - which I would have to do to fix Erik's issue.  And I am certainly not going to do that without feedback from the people who have been working on it more than the 48 hours I have.

It's not perfect.  It doesn't handle every case.  It may need to change with a later re-design to implement one of the solutions I mentioned above.  But that's what iterative development is for.  But it does stop MQTT from being completely useless at any kind of scale (at least when used with zigbee2mqtt, which uses the same availability topic for every entity - which is a valid approach).  And hell, having zigbee devices entities, like I do right now, isn't even that outrageous - and yet it's enough to break this integration completely.

But let's be clear - ANY of the solutions will probably take another week to solve (half of which is both figuring out the rest of how the MQTT module works, and deciphering the unit tests).  Which you're asking me to do, someone not a maintainer of this module, or any part of home-assistant.  I'm more inclined to just keep patching client.py on my local instance.

corporategoth replied:
Why do you think I mentioned P=NP?  There *IS* no perfect solution.

The best solution I think so far is likely delaying any subscriptions until discovery has discovered all topics to be subscribed to.  But that, as you say, has some drawbacks too.

But for now, I'm going to leave it.  I have a solution that solves the most critical problem, even if it doesn't solve everything.  I am a strong believer in "Perfect is the enemy of good."  So I guess MQTT integration will remain broken until I have enough time to dedicate to finding the "perfect" solution.  Or someone who actually knows how the code works looks at it.  Or gives me actual feedback that's not "Hmmm.. interesting .. but everything you mentioned is not perfect.  Keep banging that against the wall - you'll break through someday!"

emontnemery reviewed (changes requested):
It would be great if we could avoid subscribing to the same topic multiple times, but this PR in its current form won't work because it doesn't keep track of which messages should be replayed (only the retained ones).

Can you try to improve on that @corporategoth ?

Also, please note that the reason for us resubscribing is to make the server resend retained messages. Even though we subscribe to the same topic multiple times, an MQTT message is sent only once, and the server considers it a single subscription: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718063

Edit:
I now remember why we don't implement this.
From Section 3.3.1.3 in the [spec](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718038):
> When sending a PUBLISH Packet to a Client the Server MUST set the RETAIN flag to 1 if a message is sent as a result of a new 
 subscription being made by a Client [MQTT-3.3.1-8]. It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to a Client because it matches an established subscription regardless of how the flag was set in the message it received [MQTT-3.3.1-9].

This means we won't know if a retained message has changed, and hence I unfortunately think the solution suggested in this PR can't work.

emontnemery commented:
@corporategoth I'm marking this PR as draft. If you come up with a way where this will work, please set it to ready for review 👍

corporategoth commented:
> @corporategoth I'm marking this PR as draft. If you come up with a way where this will work, please set it to ready for review 👍

Please read my discussion on in the previous thread.  I don't think trying to solve the RETAIN problem is as critical as the rest of this change is, since this change ub-breaks HA MQTT for large number of entity scenarios.

corporategoth commented on homeassistant/components/mqtt/client.py:
```diff
@@ -503,6 +504,7 @@ async def async_subscribe(
         if not isinstance(topic, str):
             raise HomeAssistantError("Topic needs to be a string!")
 
+        already_active_subscription = self._is_active_subscription(topic)
```
Good call out - I can make that only relevant for simple topics.  Though wildcards seem fairly rare.

jbouwh replied:
Config topics are received fro wildcard subscriptions, they are not rare.

corporategoth replied:
No, I more meant the number of actual wildcard subscriptions is low, compared to simple subscriptions.  Not that the use of wildcards is rare.

In my environment, it subscribes to less than half a dozen wildcard topics, and doesn't re-subscribe to them over and over.  As opposed to the simple ones.

But as I said - this is a simple and appropriate fix.  I'll correct it.

jbouwh replied:
Subscriptions should have the configured QoS. A published value received on the subscription should be checked on a retain status. The first time it is received all processed subscriptions should be replayed initially. When new subscription requests arrive and a topic was received for it e.g. a config. It can be that some devices use retain for their config and others don't. So to be able to match replaying we should  not check the subscription topic but the topics that we have received a retained payload on, and only store those payload.
This would mean we also need to index on the topics that had retained content. We can do that when a payload is received and retained is set.
Then we should make sure to clean this all up when a topic is cleared from the broker.

corporategoth replied:
As @emontnemery mentioned - according to the spec.  the RETAIN flag for received data has no baring on what the RETAIN flag was set to when it is published.

On publish - RETAIN means "Keep this message, expiring any previously RETAINed message.  And send this message for new subscriptions."
On receive - RETAIN means "This message was sent to you because you just subscribed."
This means, after initial subscription of any topic, all future messages, whether published with RETAIN=0 or RETAIN=1, will have RETAIN=0 set in the message on the receiving side.

That's the crux of the problem Erik has with this MR, and why I mentioned 3 possible ways to think about solving that in the context of this MR. There is no way to know which messages we should store as a 'last message'.

This MR essentially treats all messages received as having been published with RETAIN=1, which is not correct, but does allow us to short circuit the necessity to send 8000+ subscription requests for the same topic.  But I need feedback on the possible 3 approaches I listed in the other thread to try and resolve this in a better way.  Each of which has downsides.  Though I also think that this MR, even as is, is such an improvement over the current status quo, the retain issue could be fixed subsequently too.

Now I could only store a message as the 'last message' if it has RETAIN=1, and then have a TTL, and if the TTL has expired, allow the subscription to go through to find out what the current RETAIN value should be - this ends up being "solution" 3 I listed above.

But QoS doesn't really affect RETAIN from what I can tell.  QOS just says  0 =  "We may not get all non-RETAIN=1 messages", 1 = "We will get every message, possibly more than once" and 2 = "We will get every message, exactly once."   The higher QoS coming with greater complexity/latency (as it requires a 0-way, 1-way or 2-way handshake).

emontnemery commented:
@corporategoth I think you're on to something with the throttling / cooldown 👍

I'd suggest to explore wrapping `homeassistant.components.mqtt.async_subscribe` with a `homeassistant.helpers.debounce.Debouncer`, with the `immediate` flag set to `False`. The debounce time can probably be relatively short, maybe 100 ms or so to get ride of the unwanted startup behavior.

corporategoth commented:
> ``.



> @corporategoth I think you're on to something with the throttling / cooldown 👍
> 
> I'd suggest to explore wrapping `homeassistant.components.mqtt.async_subscribe` with a `homeassistant.helpers.debounce.Debouncer`, with the `immediate` flag set to `False`. The debounce time can probably be relatively short, maybe 100 ms or so to get ride of the unwanted startup behavior where all entities race to setup their subscriptions.
> 
> Obviously the function `homeassistant.components.mqtt.async_subscribe` itself can't be wrapped with a Debouncer, the goal is to debounce calls to `paho.mqtt.Client.subscribe`.

Actually, I'm thinking it would be more a case of creating a list of pending subscriptions (unique by topic) - and then debounce _async_perform_subscriptions - but have that function NOT take  list of subscrtiptions, but process the list of pending subscriptions.

Since paho.mqtt.Client.subscribe can also be invoked with all topics at once - but it will be debounced automatically by virtue of the new version of _async_perform_subscriptions.

The net result would be doing a batch of subscription requests every 100 ms or so - but only ONE per topic we are waiting to subscribe for.

Combined with https://github.com/corporategoth/core/pull/2 - does that sound like a reasonable solution?

corporategoth committed:
Changed to combine and debounce subscriptions.

bdraco commented on homeassistant/components/mqtt/client.py:
```diff
@@ -547,11 +547,11 @@ def async_remove() -> None:
                 raise HomeAssistantError("Can't remove subscription twice")
             self.subscriptions.remove(subscription)
             if wildcard_topic(topic):
```
There is a conflict with a recent PR that needs to be sorted out.  It should hopefully make this a bit easier to manage this block of code as well.

corporategoth committed:
Keep non-duplication of retain messages where we can.

corporategoth committed:
Post merged

corporategoth commented:
*sigh* so it turns out I can't rely on the first message for a topic having RETAIN = 1.
So we may end up with some duplication.  However, on topics where the first message DOES come in with RETAIN=1 I can still avoid some duplication.

corporategoth commented:
Still need to fix the unit tests.

bdraco commented on homeassistant/components/mqtt/client.py:
```diff
@@ -269,15 +270,24 @@ def remove() -> None:
     return remove
 
 
-@attr.s(slots=True, frozen=True)
```
I think you could change this to a `@dataclass` instead.

bdraco commented on homeassistant/components/mqtt/client.py:
```diff
@@ -365,12 +375,23 @@ def __init__(
         self._last_subscribe = time.time()
         self._cleanup_on_unload: list[Callable[[], None]] = []
 
         self._paho_lock = asyncio.Lock()  # Prevents parallel calls to the MQTT client
+        self._subscriptions_lock = (
+            asyncio.Lock()
+        )  # Prevents parallel calls to the MQTT client
```
Can you add some comments about why two locks are needed?

corporategoth replied:
I Entities may be adding to the subscriptions list while we are performing the previous subscription request (or other MQTT operations).  I figured this would be a simple way for the subscription request to take a while, possibly, but not hold up others from adding to the list of subscriptions for next round.

But - because the processing I assume can happen on different threads, we do still need to lock the subscription list from being modified at the same time as it's being cleared by the debounced function.

bdraco commented on homeassistant/components/mqtt/client.py:
```diff
@@ -706,15 +706,10 @@ async def _async_resubscribe(self) -> None:
-                )
-            ]
-        )
+        self._pending_subscriptions = {
```
There is a race here. `self.connected` is set before this runs so its possible for another `async_subscribe` call to run before and have `self._pending_subscriptions` overwritten.

bdraco replied:
```
        self.connected = True
        dispatcher_send(self.hass, MQTT_CONNECTED)
        _LOGGER.info(
            "Connected to MQTT server %s:%s (%s)",
            self.conf[CONF_BROKER],
            self.conf[CONF_PORT],
            result_code,
        )

.....

        if (
            CONF_BIRTH_MESSAGE in self.conf
            and ATTR_TOPIC in self.conf[CONF_BIRTH_MESSAGE]
        ):

            async def publish_birth_message(birth_message: PublishMessage) -> None:
                await self._ha_started.wait()  # Wait for Home Assistant to start
                await self._discovery_cooldown()  # Wait for MQTT discovery to cool down
                await self.async_publish(
                    topic=birth_message.topic,
                    payload=birth_message.payload,
                    qos=birth_message.qos,
                    retain=birth_message.retain,
                )

            birth_message = PublishMessage(**self.conf[CONF_BIRTH_MESSAGE])
            asyncio.run_coroutine_threadsafe(
                publish_birth_message(birth_message), self.hass.loop
            )

```

I'd move this block into `_async_resubscribe` and rename it to `_async_connected`

Than there is no way that `self.connected` can be set before its run.

You'll have to change `dispatcher_send` to `async_dispatcher_send` as well dropping the `run_coroutine_threadsafe`

corporategoth committed:
Accommodating MR comments

corporategoth commented:
I could use some help with how to fix the unit tests.

The problem is that many of them expect a _mqttc.subscribe call be made - and previously there was a direct through line from call to invocation.

Now the through line is broken - it goes via. a Debouncer, which is not immediate - so includes a 100ms delay.
I'm not sure if that debouncer is even still working in the unit tests, but if it is, it's still broken the through line so I need to somehow wait for it.

I guess I could just test to make sure the debouncer itself is invoked?  But that's not as useful, because some tests are trying to do something like ensuring multiple subscription requests don't subscribe multiple times.

I'm not conversant enough with pytest and hass (and the intersection of both) to know how to fix these tests right now.  So I could use some help.

bdraco commented on homeassistant/components/mqtt/client.py:
```diff
@@ -671,16 +665,32 @@ def _mqtt_on_connect(
+                )
+            }
+
+        self.hass.async_create_task(self._subscription_debouncer.async_call())
```
I think this can move to the end of the function and be awaited instead now

corporategoth committed:
Don't bother creating an async task for subscriptions if we didn't add ie to pending subscriptions.

corporategoth committed:
Remove from pending subscriptions on unsub if the last subscriber on that topic.

corporategoth committed:
Merge branch 'dev' into prez/deduplicate_topic_subscriptions

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -594,59 +615,38 @@ def _client_unsubscribe(topic: str) -> int:
+
+        async with self._subscriptions_lock:
+            if topic in self._pending_subscriptions:
+                del self._pending_subscriptions[topic]
```
This doesn't seem right. I think we need to count the number of pending subscriptions?

emontnemery replied:
No, right above you count *active* subscription, that doesn't include *pending* subscriptions, does it?

corporategoth replied:
That's Here's the sequence:

1. A subscribes to foo
  - Subscription (class) for foo for A is created
  - pending for foo is added.
2. B subscribes to foo
  - Subscription for foo for B is created
3. A unsubscribes
  - Subscription for foo for A is removed.
  - pending for foo should NOT be removed (subscription for B exists).
4. B unsubscribes
  - Subscription for foo for B is removed.
  - pending for foo should be removed.

This can all happen before the pending subscription is processed due to the delay.
Without this code, the subscription entry for A and B would be removed.  But the pending subscription would still exist and we would still subscribe to that topic.

If I just blindly removed the pending subscription, and either B or A had not unsubscribed, we would never perform a subscription that was still needed.

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -594,59 +615,38 @@ def _client_unsubscribe(topic: str) -> int:
+    async def _async_perform_subscriptions(self) -> None:
         """Perform MQTT client subscriptions."""
 
-        # Section 3.3.1.3 in the specification:
-        # http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html
-        # When sending a PUBLISH Packet to a Client the Server MUST
-        # set the RETAIN flag to 1 if a message is sent as a result of a
-        # new subscription being made by a Client [MQTT-3.3.1-8].
-        # It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to
-        # a Client because it matches an established subscription regardless
-        # of how the flag was set in the message it received [MQTT-3.3.1-9].
-        #
-        # Since we do not know if a published value is retained we need to
-        # (re)subscribe, to ensure retained messages are replayed
```
This comment is still valid

emontnemery replied:
Please add it back

corporategoth replied:
Already done.

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -671,16 +671,32 @@ def _mqtt_on_connect(
-        self.hass.create_task(self._async_resubscribe())
+        self.hass.create_task(self._async_connect())
+
+    async def _async_connect(self) -> None:
```
Please don't move code around like this, it makes reviewing harder. The moving around can be done in a separate PR instead.

emontnemery replied:
I meant that you have renamed `_async_resubscribe` to `_async_connect` and moved it around in the class

corporategoth replied:
@emontnemery Again - at @bdraco 's suggestion.  Explicitly.  And no, I didn't just rename it.
https://github.com/home-assistant/core/pull/88072#discussion_r1109153728

_async_resubscribe was only EVER being invoked from on_connect.  And by creating _async_connect, I actually took a significant chunk of on_connect, the chunk that can/should be done async anyway, and moved it to that function, combining it with the existing async_resubscribe code (which was only being invoked by that function anyway).

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -671,16 +671,32 @@ def _mqtt_on_connect(
             return
 
-        self.connected = True
-        dispatcher_send(self.hass, MQTT_CONNECTED)
```
Why did this move?

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1292,8 +1292,9 @@ async def test_subscribe_same_topic(
 
-    When subscribing to the same topic again, SUBSCRIBE must be sent to the broker again
-    for it to resend any retained messages.
+    When subscribing to the same topic again, the last value sent to the topic
+    should be sent to the new subscriber (simulating exactly what happened with
+    the initial SUBSCRIBE call).
```
This is not correct as the retained value would be send to the new subscriber, which is not necessarily the last payload received. But I agree this is better. Again, it is not what would happen when resubscribing.

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -759,17 +759,25 @@ def _mqtt_handle_message(self, msg: mqtt.MQTTMessage) -> None:
-                ),
-            )
+
+            # Skip messages for non-wildcard topics where RETAIN=1, and we have
+            # already received a RETAIN=1 message.
+            if not _is_simple_match(subscription.topic) or not (
+                msg.retain and subscription.initialized
+            ):
+                if msg.retain:
+                    subscription.initialized = True
+                self.hass.async_run_hass_job(
+                    subscription.job,
+                    ReceiveMessage(
+                        msg.topic,
+                        payload,
+                        msg.qos,
+                        msg.retain,
+                        subscription.topic,
+                        timestamp,
+                    ),
+                )
```
Let's do the debouncing of subscribe and potential filtering of received messages in separate PRs.

emontnemery replied:
Sure, both changes are means to improve the startup performance.
However, the means are completely different.

corporategoth replied:
I have a strong belief that, when you fix a bug, you fix ALL of a bug.  In one PR.  You don't do it piecemeal.   Whether it's due to a single issue, or a combination of several.  The PR is to fix the bug, NOT to fix one piece of the bug.  If that makes the PR bigger, *shrug*.  The idea is that, you should not be able to accept one without the other, because doing so is a partial fix at best and doesn't actually fix the bug in question.

As such, since everything here is fixing the same but, even though yes, it's a contribution of two different (complementing) issues, the fix belongs in one PR.  And so no, I'm not going to split them.  If you feel strongly about it, as I said, feel free to take over the branch and split it up.

emontnemery replied:
OK, let's close this PR then.

emontnemery reviewed (commented):
Some comments

jbouwh reviewed (changes requested):
The tests still need to be addressed. For some tests we might need to remove the remove the patch on the HA MQTT client by removing the patch `mqtt_mock_entry_no_yaml_config`. In that case you need to set up the MQTT entry manually.
B.T.W. I did not test yet with the new code, but generally this is looking good.

corporategoth commented:
Ugh.  I didn't realize a bunch of my comments were stuck in pending.

corporategoth committed:
Fixed most unit tests

corporategoth commented:
I had to add a skip for the test_delayed_birth_message test.

If anyone has an idea how to fix this test, I'd be open ears.  However I'm not sure if the way this test is written will work anymore, as it assumes the subscription to the topic we publish the birth message on before we publish the birth message.  But that subscription can happen after now - since the debounce delays subscriptions by 100ms.

This is not really a production issue, as why would you otherwise subscribe to a topic you are publishing on?   But I think that's the root cause of that test hanging.  But I can't be sure.

The other tests I just fixed to accommodate the deboucne and adjusted the expected calls for subscribe to understand it will be a single call, not one per topic.

corporategoth committed:
Removed inaccurate comment

corporategoth committed:
Merge branch 'dev' into prez/deduplicate_topic_subscriptions

emontnemery commented on homeassistant/components/mqtt/client.py:
```diff
@@ -567,8 +582,14 @@ async def async_subscribe(
+                    qos > self._pending_subscriptions[topic]
+                ):
+                    self._pending_subscriptions[topic] = qos
+                    self.hass.async_create_task(
+                        self._subscription_debouncer.async_call()
+                    )
```
asyncio is single threaded, so there's no true parallelism gained by handing this off to a task.
Also, you create the task while holding the lock, but the lock is not held when the task is executing.
If that's OK, why is the task created when holding the lock?

corporategoth replied:
Note - when I say 'block' below, I mean waiting on an async call with await, not a dead sleep.

As far as I know, the task creation just puts a task in the queue for the main HASS async loop (?) to execute - more or less immediately returning.  At least that was my understanding of the function.  It doesn't hard-block it in the current flow like an await would.

The reason I want this separate is that many async_subscribe calls are going to be called back to back - as a side effect of discovery and entity creation.  Repeatedly and in sequence.  If one of them kicks off an actual subscribe call (due to the debunce time), which itself can block, then that sequence is paused until the subscribe call is completely finished.

Handing this off to HASS, means that, even in a single threaded mechanism, if a subscribe call initiates, when it blocks, it can go back to adding more subscriptions in it's sequence.  This leads to a higher chance that entities that share the same subscription topic (ie. entities on the same device) get collected into a single subscribe call, and not spread across multiple subscribe calls.  It also leads to a higher number of subscriptions being able to be done in a fewer number of calls (we can process more subscribe requests while waiting for the previous to complete).

corporategoth replied:
As for the lock - I only created it to ensure that it couldn't be modified by one entity doing a subscribe while the HASS task loop is clearing out that array, for processing it or unsub is deleting an entry from it.

If everything is single threaded, it's possible that this is not possible so the lock itself may not be necessary.  Since we don't really do anything async inherently while it's locked.  Just manipulation of a dict.

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1314,11 +1313,45 @@ async def test_subscribe_same_topic(
+async def test_subscribe_same_topic_quick(
+    hass: HomeAssistant,
+    mqtt_client_mock: MqttMockPahoClient,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
```
We should not mock the HA MQTT client here

corporategoth replied:
Don't I still need the HA MQTT mock to set the connected = True flag (line 1335)?

corporategoth replied:
That means I can't remove the HA MQTT mock - can you resolve these threads? or give me another option for all the places you asked me to remove the HA MQTT mock?

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1336,6 +1369,8 @@ async def test_not_calling_unsubscribe_with_active_subscribers(
     unsub = await mqtt.async_subscribe(hass, "test/state", record_calls)
     await mqtt.async_subscribe(hass, "test/state", record_calls)
     await hass.async_block_till_done()
+    async_fire_time_changed(hass, utcnow() + timedelta(SUBSCRIBE_COOLDOWN))
```
For this test applies the same, we are asserting on the MQTT client. We should not mock the HA MQTT client code.

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1364,20 +1399,15 @@ async def test_unsubscribe_race(
 
     async_fire_mqtt_message(hass, "test/state", "online")
     await hass.async_block_till_done()
+    async_fire_time_changed(hass, utcnow() + timedelta(SUBSCRIBE_COOLDOWN))
```
Same for this test

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1425,11 +1459,11 @@ async def test_restore_all_active_subscriptions_on_reconnect(
     await mqtt.async_subscribe(hass, "test/state", record_calls)
     await mqtt.async_subscribe(hass, "test/state", record_calls, qos=1)
     await hass.async_block_till_done()
+    async_fire_time_changed(hass, utcnow() + timedelta(SUBSCRIBE_COOLDOWN))
```
and this one

jbouwh commented on tests/components/mqtt/test_init.py:
```diff
@@ -1595,11 +1631,15 @@ async def test_subscribe_error(
     mqtt_client_mock.on_connect(mqtt_client_mock, None, None, 0)
     await hass.async_block_till_done()
-    with pytest.raises(HomeAssistantError):
+
```
and this one

emontnemery commented:
I'm closing this PR, the two suggested methods for improving startup time can be explored in separate PRs:
- Debounce subscribe calls
- Possibly do something smart with retained messages
  - What's currently in the PR is not correct, but on simple topics where we've received only a single retained message, I think we can replay that message to new subscribers instead of re-subscribing.

corporategoth commented:
@emontnemery 

*sigh* if you're going to just close the MR, despite the work I've put into it, and the amount of back and forth.  Then sod it.
I have a version of client.py that works properly, I'll not be re-submitting it.  I'll just be patching my own instance locally.  And feel bad for all the other non-programmer users.

Just so you know, until this MR or something like is is committed - HA MQTT is absolutely terrible for anyone using Zigbee2MQTT, with any reasonable sized network.  And likely for many other usages of MQTT.  And given that I'd wager zigbee2mqtt (or zwave2mqtt) is one of the most common reasons to integrate MQTT with HA, you've just decided to leave everyone using the HA MQTT integration in a broken state.  Not just with a slow startup, but most sensors being literally broken (no updates, ever) because the subscribes are dropped on the floor because of the inefficient manner subscriptions are done in the first place.

This experience, from the original erroneous closing of my tickets as 'not an issue' or 'probably zigbee2mqtts fault' without any real investigation, to needing to pull teeth to try and get any actual design feedback, to the forcing other contributors to follow YOUR specific rules for how an MR should be, has turned me off ever trying to submit any improvement again (you are the only one who cared about splitting debounce from don't re-process RETAIN=1 messages).  That might fly if you were paying people as employees, but not people who are volunteering their time and energy - trying earnestly to actually resolve a problem.

So .. in summary. You want this problem fixed now, I guess you can be a glory hound a fix it yourself.  I guess it keeps the club tight.

frenck commented:
@corporategoth Looked at it, the request was reasonable and simple, split the PRs into multiple  here: <https://github.com/home-assistant/core/pull/88072#discussion_r1109416932>

This is also documented in our developer documentation as a pre-requisite. You don't seem willing to work with that. That is up to you, the general rules of our project have been followed and if you don't want to follow those, as you answered (direct quote):

> And so no, I'm not going to split them.

We have no other option than closing if you are not willing to work with us. If you change your mind, feel free to open up a PR to improve things that complies with our general developer guidance.

Thanks 👍 

../Frenck

corporategoth commented:
@frenck As I said - I don't believe in half a fix.  Splitting the code COULD be done.  Hell, I could probably have split this into 8 MRs.  But, everything in this MR Is required to fix the issue - https://github.com/home-assistant/core/issues/87965

If I don't prevent re-processing of messages as they come in.  The issue persists.
If I don't prevent subscribing to the same topic over and over.  The issue persists.

I DID read https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible.  It states "Do not solve several issues in a single pull request." - without everything in this MR - the issue that caused this MR to be created remains unresolved.  This entire MR fixes a single issue.  It may have multiple causes, but it's still one issue.
SO, with that being said - the request to split the MR is equivalent to "Please give me 2 MRs, each with half the fix for the issue."  

But hey, your codebase.  Your rules (and decision on how strict you want to be with them).  I'll just not participate in the future.  I  know when I am running projects that people volunteer their time/effort in, I tend to cut them some slack unless it's egregious (lots of unrelated changes or something) - because well, they donated significant amounts of their time and effort.  The least I can do is spend some effort on my time (say, reviewing a slightly larger PR that fixes the whole issue) rather than making them jump through administrative hoops or rejecting all their effort because they didn't follow my process to the letter.

Coming up with this MR took me probably a good 20 hours of time (while sick with covid), probably more, to investigate the issue (and try and convince the maintainers it's not "someone else's bug") - learn the code base I had never even glanced at before (the MQTT component and some HASS), find out what the causes were, debug it.  Come up with one solution, then revise the solution on request (including benefit/drawback analysis on 3 different possible solutions), learn more of the code to figure out how the tests work, validate the fix, respond to MR feedback, etc.

Erik has had to have already reviewed the code tin its entirety to say it could be split.  So at this point, splitting it seems nit-picky and like busy work for no real gain and one major deficit (ie. a split MR will by itself fix the issue).  And just closing the MR (that was already reviewed by 3 people) is a slap in the face that says "We don't value your time and effort, all we care about is that you do exactly as we say.  Even if you have good reasons not to.  Good dog."

So .. nope.

frenck commented:
All of your ramblings above don't matter. As said, there was a review request, you can either try to help out or throw up a block, which you clearly did.

> So .. nope.

☝️ Exactly the point. Sorry to hear you don't want to contribute to our project in a way this project is handling these PRs.

> I'll just not participate in the future.

Sorry to hear that, that is up to you. All the best of luck in the future and, like said before, if you change your mind, feel free to open up a PR that will meet our requirements.


../Frenck

PS: `MR` is not a thing on GitHub 😉

corporategoth commented:
@emontnemery 

> I'm closing this PR, the two suggested methods for improving startup time can be explored in separate PRs:
>     * Debounce subscribe calls
Don't forget to actually do the smart thing, and both deduplicate topic subscribes and make SINGLE subscribe calls (ie. subscribe((topic1, qos), (topic2, qos)) as opposed to a loop of individual subscribe calls.  There is no reason to loop through each topic and subscribe.

>     * Possibly do something smart with retained messages
>       
>       * What's currently in the PR is not correct, but on simple topics where we've received only a single retained message, I think we can replay that message to new subscribers instead of re-subscribing.
What is in this PR is:
1. Deliver all messages for wildcard topics to all subscribers to those wildcard topics - regardless of RETAIN (when you subscribe to a wildcard, it can and often will send a large number of RETAIN=1 calls - one for each topic matching that wildcard).
2. Deliver all messages of RETAIN=0 messages to all current subscribers on a topic (there is also no  guarantee a topic will have ANY RETAIN=1 messages - I found this out when I tried to say "don't process retain=0 messages until after you process a retain=1 message" = it failed because the zigbee sensors don't have a persisted status at all, so no retain=1 message)
3. Deliver only one RETAIN=1 message to any simple topic subscriber, ignore all future RETAIN=1 messages for that subscriber (in this case a subscribe is an individual entity - or an individual instance of the Subscribe class)
4. We CAN send multiple subscribes on a single topic due to the debounce (say you have 20 entities that use a topic, and 5 of them get in before the debounce triggers a subscribe, and the other 15 get in the next round - the RETAIN=1 message triggered for the first subscribe will be delivered to the first 5, or possibly up to all 20, the RETAIN=1 message triggered by the next debounce subscribe will only be delivered to those who didn't get the one from the previous subscribe.  RETAIN=0 is unaffected).

This IS correct - because the documentation says that a RETAIN=1 message is received only on a new subscription, and otherwise it will send RETAIN=0 messages.  If we are receiving a RETAIN=1 message, it's because we subscribed to a topic again (presumably for a new entity) - but the existing entities already got their "baseline status" and don't need to get it/process it again.
If a message is coming in because the publisher published a new message (regardless of RETAIN=1 or 0), we will receive the message with RETAIN=0 and send it to all subscribers using this PR's logic.

Funnily enough, what you're advocating of caching and skipping subscribes is exactly my FIRST solution, which you rejected for this exact reason, because it didn't handle RETAIN=1 messages correctly.  I fixed that, and now you want to "do something smart".  As you mentioned in your FIRST reply - there is no way to know if a topic has only received a single retained message. EACH subscribe call will result in a maximum of 1 RETAIN=1 message per subscribe to that topic.  Subsequent RETAIN=1 messages (from the publishing side) will come through as RETAIN=0, and any existing subscribed topic will have no way of knowing.  So no way of realistically caching it.

I guess you could do a small optimization of if (number of received messages on a simple topic == 1, and that message was RETAIN=1) skip re-subscribe.  That would work, but have no real difference to the current code except it foregoes that subscription - which is already not as much of an issue due to debouncing / deduplicating.  The current code already avoids the exponential re-processing of RETAIN=1 messages.

But, good luck.  It's your baby now.

frenck commented:
> But, good luck. It's your baby now.

It isn't.

Locking this PR.

../Frenck
